src/main/java/com/puppycrawl/tools/checkstyle/xpath/iterators/PrecedingIterator.java

```diff
@@ -0,0 +1,88 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.xpath.iterators;
+
+import net.sf.saxon.om.AxisInfo;
+import net.sf.saxon.om.NodeInfo;
+import net.sf.saxon.tree.iter.AxisIterator;
+
+/**
+ * Recursive-free implementation of the preceding axis iterator.
+ */
+public class PrecedingIterator implements AxisIterator {
+    /**
+     * Ancestor axis iterator.
+     */
+    private final AxisIterator ancestorEnum;
+    /**
+     * Preceding-sibling axis iterator.
+     */
+    private AxisIterator siblingEnum;
+    /**
+     * Descendant axis iterator.
+     */
+    private AxisIterator descendantEnum;
+
+    /**
+     * Create an iterator over the "preceding" axis.
+     *
+     * @param start the initial context node.
+     */
+    public PrecedingIterator(NodeInfo start) {
+        ancestorEnum = start.iterateAxis(AxisInfo.ANCESTOR);
+        siblingEnum = start.iterateAxis(AxisInfo.PRECEDING_SIBLING);
+        descendantEnum = null;
+    }
+
+    /**
+     * Get the next item in the sequence.
+     *
+     * @return the next Item. If there are no more nodes, return null.
+     */
+    @Override
+    public NodeInfo next() {
+        NodeInfo result = null;
+
+        while (result == null) {
```

Comment:
done

Reply:
This is not done and may have been the one reverted. Where can't this pass coverage?

Reply:
oops, yes, this one was reverted

Reply:
> Where can't this pass coverage?

Reply:
https://timurt.github.io/checkstyle/9267/jacoco/com.puppycrawl.tools.checkstyle.xpath.iterators/PrecedingIterator.java.html#L61

Reply:
I believe the missing coverage is becasue you have 1 too many `result == null` checks.

It looks like you may get the coverage needed if you change the bottom portion to:
````
            if (result == null) {
                result = ancestorEnum.next();
                if (result != null) {
                    previousSiblingEnum = result.iterateAxis(AxisInfo.PRECEDING_SIBLING);
                }
            }
        } while (result == null);
````

if not, let me know which coverage is missing, result being `null` or not being `null`.

Reply:
> It looks like you may get the coverage needed if you change the bottom portion to:

no, we have two exit statements:
1. we reached root and there is no other node to visit, null returned
2. we found candidate node 

your code will hang, since it traversed all tree and there is no other node to visit

> if not, let me know which coverage is missing, result being null or not being null.

being null

Reply:
I see. I am not seeing how to fix it right now.